File: src/main/java/uk/laxd/dndSimulator/dice/Roll.java
package uk.laxd.dndSimulator.dice;

import java.util.Arrays;
import java.util.Collection;
import java.util.stream.Collectors;

/**
 * A roll of one or more dice, with zero or one modifiers.
 *
 * I.e. a Strength check may be represented by a Roll object, e.g.:
 *
 * 1d20 + 2
 */
public class Roll {

    private Collection<Die> dice;

    private Collection<RollModifier> modifier;

    private RollResult result;

    public Roll(Collection<Die> dice) {
        this.dice = dice;
    }

    public Roll(Die... dice) {
        this.dice = Arrays.asList(dice);
    }

    /**
     * Can be rolled once, and only once, to set the result of this roll.
     *
     * If this method is called multiple times, a {@link MultipleRollException} is thrown.
     */
    public RollResult roll() {

        // Any RollModifiers that modify the sum of the dice should be included here,
        // i.e. bless adding a d4, rage adding 2 to strength checks etc.
        // Any other modifiers that modify HOW the roll is made (i.e. advantage/disadvantage)
        // will be taken into account outside of this class


        result = new RollResult();
        result.addDice(dice.toArray(new Die[]{}));

        return result;
    }

    public void addDie(Die die) {
        this.dice.add(die);
    }

    public Collection<Die> getDice() {
        return dice;
    }

    @Override
    public String toString() {
        return dice.stream()
                .collect(Collectors.groupingBy(Die::getMaxValue)).entrySet()
                .stream()
                .map(e -> e.getValue().size() + "d" + e.getKey())
                .collect(Collectors.joining(" ")) + "(" + result + ")";
    }
}

File: app/src/main/kotlin/uk/laxd/dndSimulator/dice/Roll.kt
package uk.laxd.dndSimulator.dice

import java.util.Arrays
import java.util.stream.Collectors

/**
 * A roll of one or more dice, with zero or one modifiers.
 *
 * I.e. a Strength check may be represented by a Roll object, e.g.:
 *
 * 1d20 + 2
 */
open class Roll(val dice: MutableCollection<Die>, val modifier: Int = 0) {
    private var result: RollResult? = null

    constructor(vararg dice: Die, modifier: Int = 0) : this(mutableListOf(*dice), modifier)

    /**
     * Creates a random number between 1 + [modifier] and [Die.maxValue] + [modifier]
     */
    open fun roll(): RollResult {
        // Any RollModifiers that modify the sum of the dice should be included here,
        // i.e. bless adding a d4, rage adding 2 to strength checks etc.
        // Any other modifiers that modify HOW the roll is made (i.e. advantage/disadvantage)
        // will be taken into account outside of this class
        val result = RollResult()

        for(die in dice) {
            result.addDice(die)
        }

        result.addModifier(modifier)

        return result
    }

    fun addDie(die: Die) {
        this.dice.add(die)
    }

    override fun toString(): String {
        return this.dice.stream()
                .collect(Collectors.groupingBy(Die::maxValue)).entries
                .stream()
                .map { e: Map.Entry<Int, List<Die?>> -> e.value.size.toString() + "d" + e.key }
                .collect(Collectors.joining(" ")).toString() + "(" + result + ")"
    }
}

class AdvantageRoll(val roll: Roll): Roll(roll.dice, roll.modifier) {
    override fun roll(): RollResult {
        return listOf(super.roll(), super.roll())
            .maxByOrNull { r -> r.outcome }!!
    }

    override fun toString(): String {
        return "AdvantageRoll(roll=$roll)"
    }

}

class DisadvantageRoll(val roll: Roll): Roll(roll.dice, roll.modifier) {
    override fun roll(): RollResult {
        return listOf(super.roll(), super.roll())
            .minByOrNull { r -> r.outcome }!!
    }
}